File: Data Structures & Algorithms/remove-element/submission-4.kt
class Solution {
    fun removeElement(nums: IntArray, `val`: Int): Int {
        var left = 0
        var right = nums.lastIndex
        while (left <= right) {
            if (nums[left] != `val`) {
                left += 1
                continue
            }
            while (left < right && nums[right] == `val`) {
                right -= 1
            }
            nums[left] = nums[right]
            right -= 1
        }
        return left
    }
}

File: Data Structures & Algorithms/kth-largest-element-in-an-array/submission-4.swift
class Solution {
    func findKthLargest(_ nums: [Int], _ k: Int) -> Int {

        var mutableNums = nums

        func swap(_ index: Int, _ index1: Int) {
            let temp = mutableNums[index]
            mutableNums[index] = mutableNums[index1]
            mutableNums[index1] = temp
        }

        var output: Int? = nil

        func quickSort(_ left: Int, _ right: Int) {

            if left > right {
                return
            }

            if output != nil {
                return
            }

            let pivotIndex = Int.random(in: left ... right)
            let pivotValue = mutableNums[pivotIndex]

            swap(pivotIndex, right)

            var start = left
            for i in left ..< right {
                if mutableNums[i] < pivotValue {
                    swap(start, i)
                    start += 1
                }
            }

            swap(start, right)

            if start == mutableNums.count - k {
                output = mutableNums[start]
                return
            }
            quickSort(left, start - 1)
            quickSort(start + 1, right)
        }

        quickSort(0, mutableNums.count - 1)

        return output ?? -1

    }
}
